cpp/src/Ice/ConnectionI.cpp

```diff
@@ -1245,6 +1240,8 @@ Ice::ConnectionI::message(ThreadPoolCurrent& current)
             // - the read operation on the transport can't continue without blocking
             if (current.operation & SocketOperationRead)
             {
+                assert(_readHeader || !isAtRest());
```

Comment:
> Is it possible for isAtRest() to have changed since line 1220 - all within the same mutex lock?

It's not, I fixed it to use `wasAtRest`.

> Also, it would be nice to add a comment explaining the logic of this assert. I think it's because a ValidateConnection frame is only 14 bytes and as a result doesn't require reading more bytes.

Added: 
```
// If not expecting to read a header, we're expecting to read the reminder of a message and in this case
// the connection shouldn't be at rest.
```